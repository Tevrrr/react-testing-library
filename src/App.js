/** @format */

import { useState, useEffect } from 'react';
import {  Route, Routes, Link } from 'react-router-dom';
import AppRouter from './router/AppRouter';

import Users from './users/Users';

function App() {
	const [data, setData] = useState(null);
	const [toggle, setToggle] = useState(false);
	const [value, setValue] = useState('');
	const onClick = () => {
		setToggle(!toggle);
	};

	useEffect(() => {
		setTimeout(() => {
			setData('Hello World 2');
		}, 1000);
	}, []);

	return (
		<div className='App'>
			<h1>Hello World</h1>
			{data && <h2 style={{ color: 'red' }}>{data}</h2>}
			<input
				value={value}
				onChange={(e) => setValue(e.target.value)}
				type='text'
				placeholder='input...'
			/>
			<p data-testid='input-value'>{value}</p>
			<button data-testid='toggle-btn' onClick={onClick}>
				{' '}
				BUTTON
			</button>
			{toggle && <h3 data-testid='toggle-item'>Hello World 3</h3>}
			<Users />
			<p>
				<Link data-testid='main-link' to={'Main'}>
					Main
				</Link>{' '}
				<Link data-testid='about-link' to={'About'}>
					About
				</Link>
			</p>

			<AppRouter/>
		</div>
	);
}

export default App;
