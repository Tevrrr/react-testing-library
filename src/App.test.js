/** @format */

import { toBeInTheDocument } from '@testing-library/jest-dom/dist/matchers';
import { render, screen, fireEvent } from '@testing-library/react';
import  userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router';
import App from './App';

describe('Test APP', () => {
	// find  - асинхронное использование
	// get   - поиск какого-то компонента иначе будет ошибка
	// query - проверка на отсутствие компонента или его наличие неважно
	test('get', () => {
		render(<MemoryRouter><App /></MemoryRouter>);
		const helloWorldElement = screen.getByText(/hello world/i);
		const btn = screen.getByRole('button');
		const input = screen.getByPlaceholderText(/input.../i);
		expect(helloWorldElement).toBeInTheDocument();
		expect(btn).toBeInTheDocument();
		expect(input).toBeInTheDocument();
		expect(input).toMatchSnapshot();
		// screen.debug(); Скрин отрендереного компонента
	});

	test('query', () => {
		render(
			<MemoryRouter>
				<App />
			</MemoryRouter>
		);
		const noHelloWorldElement = screen.queryByText(/no hello world/i);
		expect(noHelloWorldElement).toBeNull();
    });
    
	// test('find', async () => {
	// 	render(
	// 		<MemoryRouter>
	// 			<App />
	// 		</MemoryRouter>
	// 	);
	// 	screen.debug();
	// 	const helloWorld2Element = await screen.findByText(/hello world 2/i);
	// 	expect(helloWorld2Element).toBeInTheDocument();
	// 	expect(helloWorld2Element).toHaveStyle({color: 'red'});
	// 	screen.debug();
    // });
    
    test('click event',  () => {
        render(
			<MemoryRouter>
				<App />
			</MemoryRouter>
		);
        const btnElement = screen.getByTestId('toggle-btn');
        expect(screen.queryByTestId('toggle-item')).toBeNull();

        fireEvent.click(btnElement)
        expect(screen.getByTestId('toggle-item')).toBeInTheDocument();

        fireEvent.click(btnElement);
		expect(screen.queryByTestId('toggle-item')).toBeNull();
    });

    test('input event', () => {
		render(
			<MemoryRouter>
				<App />
			</MemoryRouter>
		);
		const input = screen.getByPlaceholderText(/input.../i);
		expect(screen.queryByTestId('input-value')).toContainHTML('')
        fireEvent.input(input, {
            target: { value: 'Hello!'}
        })
		expect(screen.queryByTestId('input-value')).toContainHTML('Hello!');

    });
    
    test('input event userEvent', () => {
		render(
			<MemoryRouter>
				<App />
			</MemoryRouter>
		);
		const input = screen.getByPlaceholderText(/input.../i);
		expect(screen.queryByTestId('input-value')).toContainHTML('');
		userEvent.type(input, 'Hello2!');
		expect(screen.queryByTestId('input-value')).toContainHTML('Hello2!');
	});
});
