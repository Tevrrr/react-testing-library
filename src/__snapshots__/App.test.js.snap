// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test APP get 1`] = `
<input
  placeholder="input..."
  type="text"
  value=""
/>
`;
